lightly/loss/emp_ssl_loss.py

```diff
@@ -0,0 +1,111 @@
+"""Code for EMP-SSL Loss, largely taken from https://github.com/tsb0601/EMP-SSL"""
+
+from typing import List, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch import Tensor
+
+
+class SimilarityLoss(nn.Module):
+    """Loss representing the similiarity between the patch embeddings and the average of the patch embeddings.
+
+    Args:
+        z_list:
+            List of patch embeddings.
+    Returns:
+        Similarity loss.
+
+    """
+
+    def __init__(
+        self,
+    ) -> None:
+        super().__init__()
+
+    def forward(self, z_list: List[Tensor]) -> Tensor:
+        num_patch = len(z_list)
+        z_stacked = torch.stack(z_list, dim=0)
+        z_avg = z_stacked.mean(dim=0)
+
+        z_sim = torch.tensor(0.0, device=z_stacked.device, requires_grad=True)
+        for i in range(num_patch):
+            z_sim = z_sim + F.cosine_similarity(z_stacked[i], z_avg, dim=1).mean()
+
+        z_sim = z_sim / num_patch
+
+        return -z_sim
+
+
+class TotalCodingRate(nn.Module):
+    def __init__(self, eps: float = 0.2) -> None:
+        super(TotalCodingRate, self).__init__()
+        self.eps = eps
+
+    def compute_discrimn_loss(self, W: Tensor) -> Tensor:
+        """Discriminative Loss."""
+        d, b = W.shape
+        I = torch.eye(d, device=W.device)
+        scalar = d / (b * self.eps)
```

Comment:
I'll mention that, and I think it's fine to keep it as just epsilon instead of performing a redundant multiplication each time the loss is calculated.